src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/ECDsaOpenSsl.cs

```diff
@@ -265,107 +295,122 @@ private static void CopySignatureField(byte[] signatureField, byte[] response, i
+                throw new PlatformNotSupportedException(string.Format(SR.Cryptography_InvalidCurve, oid));
 
-        private struct SupportedAlgorithm
-        {
-            public SupportedAlgorithm(int keySize, int nid)
-                : this()
+            if (!Interop.Crypto.EcKeyGenerateKey(key))
+                throw Interop.Crypto.CreateOpenSslCryptographicException();
+
+            // Set the key size for current curve
+            if (_curve != null)
             {
-                KeySize = keySize;
-                Nid = nid;
+                ForceSetKeySize(Interop.Crypto.EcKeyGetSize(key));
+                _curve = null; // The key has been generated
             }
 
-            public int KeySize { get; private set; }
-            public int Nid { get; private set; }
+            return key;
         }
 
-        private static readonly SupportedAlgorithm[] s_supportedAlgorithms =
-            RemoveAlgorithmsUnsupportedByOs(
-                new SupportedAlgorithm[]
-                {
-                    new SupportedAlgorithm(keySize: 224, nid: Interop.Crypto.NID_secp224r1),
-                    new SupportedAlgorithm(keySize: 256, nid: Interop.Crypto.NID_X9_62_prime256v1),
-                    new SupportedAlgorithm(keySize: 384, nid: Interop.Crypto.NID_secp384r1),
-                    new SupportedAlgorithm(keySize: 521, nid: Interop.Crypto.NID_secp521r1),
-                }
-            );
-
-        private static SupportedAlgorithm[] RemoveAlgorithmsUnsupportedByOs(SupportedAlgorithm[] supportedAlgorithms)
+        private void FreeKey()
         {
-            List<SupportedAlgorithm> filteredSupportedAlgorithms = new List<SupportedAlgorithm>(supportedAlgorithms.Length);
-            foreach (SupportedAlgorithm supportedAlgorithm in supportedAlgorithms)
+            if (_key != null)
             {
-                int nid = supportedAlgorithm.Nid;
-                using (SafeEcKeyHandle key = Interop.Crypto.EcKeyCreateByCurveName(nid))
+                if (_key.IsValueCreated)
                 {
-                    if (key != null && !key.IsInvalid)
-                    {
-                        filteredSupportedAlgorithms.Add(supportedAlgorithm);
-                    }
+                    SafeEcKeyHandle handle = _key.Value;
+                    if (handle != null)
+                        handle.Dispose();
                 }
+                _key = null;
             }
-            return filteredSupportedAlgorithms.ToArray();
+        }
+
+        private void SetKey(SafeEcKeyHandle newKey)
+        {
+            // Use ForceSet instead of the property setter to ensure that LegalKeySizes doesn't interfere
+            // with the already loaded key.
+            ForceSetKeySize(Interop.Crypto.EcKeyGetSize(newKey));
+
+            _key = new Lazy<SafeEcKeyHandle>(() => newKey);
+
+            // Have Lazy<T> consider the key to be loaded
+            var dummy = _key.Value;
```

Comment:
I should have had a todo -- it used to support lazy but when generatekey was added it was easier to just hydrate it always pending the perf discussion. Since we had that discussion, I will remove the laziness. Thanks